File: DataStructures/list.c
#include <stdio.h>
#include <stdlib.h>


struct list{
    int val;
    struct list *next;
};

struct list *insert(struct list *firstNode, int x) {

    struct list *newNode = (struct list*)malloc(sizeof(struct list));
    newNode->val = x;
    newNode->next = NULL;
    if(!firstNode)
        return newNode;
    else if (firstNode->val > x)
    {
        newNode->next = firstNode;
        return newNode;
    } else {
        struct list *currentNode = firstNode;
        struct list *nextNode = firstNode->next;
        while (nextNode != NULL && nextNode->val < x) {
            currentNode = nextNode;
            nextNode = nextNode->next;
        }
        newNode->next = nextNode;
        currentNode->next = newNode;
        return firstNode;
    }

}

struct list *search(struct list *firstNode, int x) {
    struct list *currentNode = firstNode;
    while(currentNode->val < x && currentNode->next != NULL )
        currentNode = currentNode ->next;
    if (currentNode->val == x)
        return currentNode;
    else
        return NULL;
}

void show(struct list *firstNode) {
    struct list *currentNode = firstNode;
    while(currentNode != NULL)
    {
        printf("%d  ",currentNode->val);
        currentNode = currentNode ->next;
    }
}

int max (struct list *firstNode) {

    struct list *currentNode = firstNode;
    while(currentNode->next != NULL )
        currentNode = currentNode ->next;
    return currentNode->val;
}

void deleteList(struct list *firstNode) {
    struct list *currentNode = firstNode;
    while(currentNode != NULL)
    {
        firstNode = currentNode -> next;
        free(currentNode);
        currentNode = firstNode;
    }
}

int length (struct list* node){
    if(node == NULL)
        return 0;
    else
        return length(node->next) + 1;
}

int main(){
    FILE *in = fopen("data.txt", "r");
    int x;
    struct list *List = NULL;
    while (fscanf(in, "%d", &x) != EOF)
        List = insert(List, x);
    fclose(in);
    printf("The biggest element: %d \n", max(List));
    printf("Next value after 21 is %d \n", search(List,21)->next->val);
    printf("Length of list is %d \n", length(List));
    printf("All elements: ");
    show(List);
    deleteList(List);
}

File: DataStructures/list.py
class Node:
    def __init__(self, val=None):
        self.next = None
        self.val = val

    def __del__(self):
        if not (self.next is None):
            del self.next
        print("Delete " + str(self.val));

    def insert(self, val):
        if self.val is None:
            self.val = val
            return self
        elif self.val > val:
            new_node = Node(val)
            new_node.next = self
            return newNode
        else:
            new_node = Node(val)
            current_node = self
            next_node = self.next
            while not(next_node is None) and next_node.val < val:
                current_node = next_node
                next_node = next_node.next
            new_node.next = next_node
            current_node.next = new_node
            return self

    def search(self, val):
        current_node = self
        while val > current_node.val and not(current_node.next is None):
            current_node = current_node.next
        if val == current_node.val:
            return current_node
        else:
            return None

    def show(self):
        print(self.val)
        if not (self.next is None):
            self.next.show()

    def length(self):
        if self.next is None:
            return 1
        else:
            return self.next.length() + 1

    def max(self):
        if not (self.next is None):
            return self.next.max()
        return self.val


list = Node()
with open("data.txt") as input:
    data = input.read().splitlines()
for x in data:
        list = list.insert(int(x))
print("Length of BST: " + str(list.length()))
print("The biggest element: " + str(list.max()))
print("Next value after 21 is " + str(list.search(21).next.val))
list.show()
del list
